### PR TITLE
JDK-8234682: The order of @param in the generated docs should match the method signature

### PR DESCRIPTION
**Pull request by hns**
Please review a change in how documentation from `@param` tags is generated. 

The old code generates parameter documentation for each `@param` in the order in which the tags occur in the comment, then adds documentation from inherited `@param` tags for undocumented parameters. 

The new code always generates documentation in the order in which actual parameters are declared in the code, using local or inherited `@param` tags as appropriate. Any `@param` tags that do not have a matching parameter are added afterwards.

Note that `@param` is not just used for parameters of executable members but also type parameters and record components. The second commit of this PR fixes a `ClassCastException` for these uses that was caused by the first commit and adds a few tests for it.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8234682](https://bugs.openjdk.java.net/browse/JDK-8234682): The order of @param in the generated docs should match the method signature

### Reviewers
 * [Pavel Rappo](https://openjdk.java.net/census#prappo) (@pavelrappo - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7046/head:pull/7046` \
`$ git checkout pull/7046`

Update a local copy of the PR: \
`$ git checkout pull/7046` \
`$ git pull https://git.openjdk.java.net/jdk pull/7046/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7046`

View PR using the GUI difftool: \
`$ git pr show -t 7046`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7046.diff">https://git.openjdk.java.net/jdk/pull/7046.diff</a>

</details>

### CODE REVIEW

**Comment by RogerRiggs** [comment]
It would be helpful if the order of the \@param tags in the source was matched the declaration.
Can a warning be generated if the \@param tags are out of order?

**Comment by hns** [comment]
> It would be helpful if the order of the @param tags in the source was matched the declaration. Can a warning be generated if the @param tags are out of order?

I think it is worth considering. Does it have to be a warning or would an informational notice also do? For something to be reported as warning it should have a certain gravity, I'm not sure `@param` order qualifies for it.

**Comment by jonathan-gibbons** [comment]
> It would be helpful if the order of the @param tags in the source was matched the declaration.
> Can a warning be generated if the @param tags are out of order?

That seems like a good suggestion, for a different PR.  The obvious place to implement such a warning would be in DocLint.

**Comment by jonathan-gibbons** [comment]
> > It would be helpful if the order of the @param tags in the source was matched the declaration. Can a warning be generated if the @param tags are out of order?
> 
> I think it is worth considering. Does it have to be a warning or would an informational notice also do? For something to be reported as warning it should have a certain gravity, I'm not sure `@param` order qualifies for it.

That's why it should be an opt-in lint-style warning. It's every bit as useful as other lint warnings, like javac's "extra semi-colon" and other minor grumbles.

**Comment by hns** [comment]
> > It would be helpful if the order of the @param tags in the source was matched the declaration.
> > Can a warning be generated if the @param tags are out of order?
> 
> That seems like a good suggestion, for a different PR. The obvious place to implement such a warning would be in DocLint.

I filed an issue: https://bugs.openjdk.java.net/browse/JDK-8279931

**Comment by pavelrappo** [comment]
Out of curiosity, how much of a churn does this PR introduce into the generated JDK API documentation? In any case, we should check it, to make sure that nothing surprising comes up.

**Comment by hns** [comment]
> Out of curiosity, how much of a churn does this PR introduce into the generated JDK API documentation? In any case, we should check it, to make sure that nothing surprising comes up.

That's a very good question! Turns out there are more than 50 classes in JDK that contain `@param` tags in wrong order. For instance, there is a [Collectors.groupingBy][1] method that gets both type and method parameter order wrong. I'm surprised not more of these get reported because I think it's quite confusing. 

[1]: https://download.java.net/java/early_access/jdk18/docs/api/java.base/java/util/stream/Collectors.html#groupingBy(java.util.function.Function,java.util.function.Supplier,java.util.stream.Collector)

I checked a few cases against the documentation generated with this PR and the it fixes the problem.

File: test/langtools/jdk/javadoc/doclet/testParamTaglet/pkg/C.java [hunk 1]
```diff
@@ -38,4 +38,39 @@ public void regularParams(int param1, int param2) {}
...
+    /**
+     * Overriding method with missing/additional/duplicate param tags in non-declaration order.
+     *
+     * @param x does not exist
+     * @param b a boolean
+     * @param i an int
+     * @param x duplicate
+     * @param b another duplicate
+     */
+    public void unorderedParams(int i, double d, boolean b) {}
```
Comment (pavelrappo) [inline test/langtools/jdk/javadoc/doclet/testParamTaglet/pkg/C.java:50]: Consider adding `@Override` to this method and to `nonMatchingParams`.

File: test/langtools/jdk/javadoc/doclet/testParamTaglet/TestParamTaglet.java [hunk 2]
```diff
@@ -48,22 +48,51 @@ public void test() {
...
+                // Regular param tags.
                 """
                     <dt>Parameters:</dt>
                     <dd><code>param1</code> - testing 1 2 3.</dd>
-                    <dd><code>param2</code> - testing 1 2 3.""",
-                //Param tags that don't match with any real parameters.
+                    <dd><code>param2</code> - testing 1 2 3.</dd>
+                    </dl>""",
+                // Param tags that don't match with any real parameters.
+                // {@inherit} doc misuse does not cause doclet to throw exception.
```
Comment (pavelrappo) [inline test/langtools/jdk/javadoc/doclet/testParamTaglet/TestParamTaglet.java:63]: Should we change this to `{@inheritDoc}`?

File: test/langtools/jdk/javadoc/doclet/testParamTaglet/pkg/C.java [hunk 3]
```diff
@@ -38,4 +38,39 @@ public void regularParams(int param1, int param2) {}
...
+
+    /**
+     * Generic method with mixed/missing param tags.
+     *
+     * @param t1 param 1
+     * @param <T2> type 2
+     * @param t3 param 3
+     */
+    public static <T1, T2, T3> void genericMethod(T1 t1, T2 t2, T3 t3) {
```
Comment (pavelrappo) [inline test/langtools/jdk/javadoc/doclet/testParamTaglet/pkg/C.java:59]: You use `{}` to denote an empty body everywhere else in this file. Is there a reason you split it into two lines here?

File: test/langtools/jdk/javadoc/doclet/testParamTaglet/pkg/C.java (hunk 1 above)
Comment (hns) [reply test/langtools/jdk/javadoc/doclet/testParamTaglet/pkg/C.java:52]: Thanks for the feedback, Pavel. I added a commit that addresses all your suggestions.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/ParamTaglet.java [hunk 4]
```diff
@@ -225,23 +200,15 @@ private Content getInheritedTagletOutput(ParamKind kind, Element holder,
      *
      * @param paramTags the array of {@code @param DocTree} to convert.
      * @param writer the TagletWriter that will write this tag.
-     * @param alreadyDocumented the set of exceptions that have already
-     *        been documented.
-     * @param rankMap a {@link java.util.Map} which holds ordering
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/ParamTaglet.java:230]: It's ironical that a method whose job, at least in part, is to warn about duplicated `@param` tags had one of its `@param` tags duplicated. Was it an easter egg of sorts?

**Review by pavelrappo — APPROVED** [review]
1. As you pointed elsewhere, `rank` is actually a position of a particular parameter. Later, we could rename `rank` into `position` or something similarly appropriate.
2. Although "ranks" precede this patch, my gut feeling is that rank being String rather than Integer, is a bad idea. But as long as "ranks" are not compared for order (only for equality), we are okay; alphanumeric order differs from that of numeric: "1", "11", "2" vs 1, 2, 11. We should revisit "ranks" later.

**Pull request closed without merging**